e2e_playwright/st_form_container_association.py

```diff
@@ -0,0 +1,95 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import streamlit as st
+
+# Element created in a form located in the sidebar.
+form_0_key = "form_0"
+with st.sidebar.form(form_0_key):
+    value = st.checkbox(f"in {form_0_key}")
+    st.form_submit_button(f"{form_0_key} submit")
+st.sidebar.write(f"{form_0_key} value:", value)
+
+# Element created in the sidebar, outside the form.
+form_1_key = "form_1"
+with st.form(form_1_key):
+    "Empty Form 1"
+    value = st.sidebar.checkbox(f"NOT in {form_1_key}")
+    st.form_submit_button(f"{form_1_key} submit")
+st.write(f"{form_1_key} value:", value)
+
+# Parent block created outside a form; element created inside a form.
+form_2_key = "form_2"
+cols = st.columns(2)
+with st.form(form_2_key):
+    "Empty Form 2"
+    value = cols[0].checkbox(f"NOT in {form_2_key}")
+    st.form_submit_button(f"{form_2_key} submit")
+st.write(f"{form_2_key} value:", value)
+
+# Parent block and element created inside a form.
+form_3_key = "form_3"
+with st.form(form_3_key):
+    cols = st.columns(2)
+    with cols[0]:
+        value = st.checkbox(f"in {form_3_key}")
+    st.form_submit_button(f"{form_3_key} submit")
+st.write(f"{form_3_key} value:", value)
+
+# Parent block created inside a form; element created outside a form.
+form_4_key = "form_4"
+with st.form(form_4_key):
+    cols = st.columns(2)
+    st.form_submit_button(f"{form_4_key} submit")
+value = cols[0].checkbox(f"in {form_4_key}")
+st.write(f"{form_4_key} value:", value)
+
+# DG created outside a form; element created inside a form.
+form_5_key = "form_5"
+empty = st.empty()
+with st.form(form_5_key):
+    "Empty Form 5"
+    value = empty.checkbox(f"NOT in {form_5_key}")
+    st.form_submit_button(f"{form_5_key} submit")
+st.write(f"{form_5_key} value:", value)
```

Comment:
for my own learning, what's the exact difference between this and the `form_2` case [here](https://github.com/streamlit/streamlit/pull/8774/files#diff-ef39285c745e265afa0b847f7c75fadacb90f886277a7a65dff14efb847abc93R32) (basically boiling down to what's the exact difference between `st.columns` and `st.empty` concept-wise ?

Reply:
The script code for this test is mostly adapted from the old script from Tim. In general, `st.empty` is a single-element placeholder, while `st.columns` works on a multi-element container.. so there is at least some difference. But not sure if this is strictly required to test both cases. But doing this doesn't add a lot of overhead in this case.

Reply:
ah I see yeah that makes sense. Thank you!